client/src/main/java/io/pravega/client/control/impl/Controller.java

```diff
@@ -364,6 +365,14 @@
      */
     CompletableFuture<Transaction.Status> checkTransactionStatus(final Stream stream, final UUID txId);
 
+    /**
+     * Gets an {@link AsyncIterator} on TxnResponse in stream.
```

Comment:
I see an internal client API being added, but it's not used anywhere or public. So, what is it for?

Reply:
**listCompletedTransactions** is public. Currently we only have a **checkTransactionState** API that returns the state for a single transaction. So user need to invoke **checkTransactionState** api multiple time to check multiple transaction status. This causes a flood of checkTransactionState API calls on Controller often causing it to go OOM. **listCompletedTransactions** is optimized way to check status of multiple transactions.

Reply:
> listCompletedTransactions is public.

This interface is in a ".impl" directory which indicates it is for internal use and hence is stripped from all public documentation and not considered part of our API. So I would not consider this 'public'.

Reply:
>  So user need to invoke checkTransactionState api multiple time to check multiple transaction status. This causes a flood of checkTransactionState API calls on Controller often causing it to go OOM.

If these calls are async the first part doesn't sound like a problem. If there is an OOM we need to fix that first, as that is a bug. Is there an issue on it?

Reply:
> listCompletedTransactions is optimized way to check status of multiple transactions

I'm not sure that is true. If I want to check the status of 10 transactions how is it faster to iterate over all the transactions that have ever occurred rather than just asking about those 10?